text/0000-reseeding-rand.md

```diff
@@ -0,0 +1,682 @@
+
+```rust
+fn thread_rng() -> ThreadRng {
+    // fetch the cached RNG from TLS, or create it if it doesn't exist
+}
+
+fn random<T: Random<D>, D>(data: D) -> T {
+    thread_rng().gen(data)
+}
+```
+
+### Alternatives
+
+There is valid concern that the choice of RNGs described below leaves
+users looking at jargony names for RNGs, which no particularly obvious
+default choice; possibly leading to poor selection. Functions like:
+
+```rust
+fn secure_rng() -> OsRng { ... }
+fn balanced_rng() -> ChaChaRng { ... }
+fn weak_rng() -> XorShiftRng { ... }
+```
+
+could be provided to give some guidance. As a counterpoint, the
+`collections` library does not provide `ordered_map() -> BTreeMap`,
+`unordered_map() -> HashMap`.
+
+## Error handling
+
+[error handling]: #error-handling
+
+As it stands, there is no error handling at all, which is
+unfortunate. Most PRNGs do not have any error conditions (some define
+reaching the end of the sequence and repeating as an error), but the
+interface with the operating system may.
+
+Reconciling this is similar to IO error handling, where some readers
+and writers are guaranteed to not return errors (e.g. reading from
+memory). This consideration is postponed until the design in
+[RFC 576](https://github.com/rust-lang/rfcs/pull/576) is definite, but
+it would likely be along the lines of making each function return
+`Result<..., <Self as Reader>::Error>`. instead of just `...`. This
+may have unfortunate ergonomic impact, and may be more complicated if
+building on `Iterator<u32>`.
+
+## Provided RNGs
+
+[provided rngs]: #provided-rngs
+
+The module will provide 3 basic algorithms to try to provide more clarity:
+
+- `OsRng`, a "true" random number generator calling into the operating system,
+- `XorshiftRng`, an [xorshift][xorshift] random number generator (or
+  one of the variations), for extremely high performance random
+  numbers,
+- `ChaChaRng`, a [ChaCha] random number generator, for "secure"
+  user-space random numbers.
+
+[xorshift]: http://en.wikipedia.org/wiki/Xorshift
+
+### Algorithms
+
+#### `OsRng`
+
+An *unbuffered* wrapper around the best available operating system
+primitive, e.g.
+
+- `CryptGenRandom` on Windows,
+- `getrandom(2)` on recent Linux versions,
+- `/dev/urandom` on OSX, BSD, and older linux versions.
+
+(See [the current `std::rand`][current] docs for more info and
+discussion about this, particularly `/dev/urandom`. NB. to future
+readers, the information will still be available in the docs after
+this RFC lands, but may have moved.)
+
+[current]: http://doc.rust-lang.org/nightly/std/rand/#cryptographic-security
+
+The lack of buffering by default means every call will interact with
+the operating system, making it very slow; but this is the only way
+that the RFC author is comfortable to call it cryptographically
+secure, adding complications is risky e.g. naively buffering would not
+be cryptographically secure due to `fork` sharing memory.
```

Comment:
Existing code isn't going to be calling Rust in between `fork` and `exec`. It's unsafe to call any code not explicitly designed for forking there, and generally any multithreaded code is too hard to get working.

It's pretty easy to solve this problem for a global random-number generator via `pthread_atfork` but there's a high cost to trying to solve it elsewhere, and you're going to get lots of deadlocks and memory corruption from misusing Rust code for an environment it wasn't designed for anyway. Rust certainly has no correctness guarantees when you totally break it like this.